sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,383 @@
+import json
+import os
+import os.path
+
+import boto3
+from botocore.errorfactory import ClientError
+
+from sacred.commandline_options import CommandLineOption
+from sacred.dependencies import get_digest
+from sacred.observers.base import RunObserver
+from sacred.serializer import flatten
+import re
+import socket
+
+DEFAULT_S3_PRIORITY = 20
+
+
+def _is_valid_bucket(bucket_name):
+    # See https://docs.aws.amazon.com/awscloudtrail/latest/userguide/
+    # cloudtrail-s3-bucket-naming-requirements.html
+    if len(bucket_name) < 3 or len(bucket_name) > 63:
+        return False
+
+    labels = bucket_name.split(".")
+    # A bucket name consists of "labels" separated by periods
+    for label in labels:
+        if len(label) == 0 or label[0] == "-" or label[-1] == "-":
+            # Labels must be of nonzero length,
+            # and cannot begin or end with a hyphen
+            return False
+        for char in label:
+            # Labels can only contain digits, lowercase letters, or hyphens.
+            # Anything else will fail here
+            if not (char.isdigit() or char.islower() or char == "-"):
+                return False
+    try:
+        # If a name is a valid IP address, it cannot be a bucket name
+        socket.inet_aton(bucket_name)
+    except socket.error:
+        return True
+
+
+def s3_join(*args):
+    return "/".join(args)
+
+
+class S3Observer(RunObserver):
+    VERSION = "S3Observer-0.1.0"
+
+    @classmethod
```

Comment:
We have been discussing to remove the factory methods from the observers. I think in this case you could just move the two lines of logic into the `__init__` method, as `create` does not contain much logic.

Reply:
Sounds good! I was only using the `create` structure because I wanted to be consistent with the pattern used in existing observers, so I'd be happy to go back to the simpler `__init__` approach.